src/validation.ts

```diff
@@ -0,0 +1,123 @@
+import { Client, cacheExchange, fetchExchange } from '@urql/core';
+import {
+  type DscvrFramesRequest,
+  type DscvrValidationResponse,
+  dscvrClientProtocolPrefix,
+  DscvrUntrustedData,
+} from './types';
+import { DEFAULT_DSCVR_API_URL } from './default';
+
+type ValidationType = string | number | bigint | null | undefined;
+interface ValidatedQueryResult {
+  user: {
+    id: string;
+  };
+  content?: {
+    id: string;
+  };
+  state: string | null;
+  inputText: string | null;
+  timestamp: string;
+  url: string;
+  buttonIndex: number;
+}
+
+const validationQuery = `#graphql
+    query validate($message: String!) {
+      unpackFrameMessage(message: $message) {
+        user {
+            id
+        },
+        content {
+            id
+        }
+        buttonIndex,
+        state,
+        url,
+        timestamp,
+        inputText,
+      }
+    }
+`;
+
+export const validateClientProtocol = (clientProtocol: string) => {
+  return clientProtocol.startsWith(dscvrClientProtocolPrefix);
+};
+
+const validateField = (
+  validated: ValidationType,
+  untrusted: ValidationType
+): boolean => {
+  return (!validated && !untrusted) || validated === untrusted;
+};
+
+const validateData = (
+  validatedResult: DscvrValidationResponse,
+  untrustedData: DscvrUntrustedData
+): boolean => {
+  const untrustedDataTransformed: DscvrUntrustedData = {
+    ...untrustedData,
+    timestamp: Number(untrustedData.timestamp), // TODO: remove this when Proxy is fixed
+  };
+
+  const isValidKey = (k: string): k is keyof DscvrValidationResponse =>
+    k in validatedResult;
+
+  const untrustedDataKeys = Object.keys(untrustedDataTransformed);
+  const isValid = untrustedDataKeys.reduce((valid, key) => {
+    if (isValidKey(key)) {
+      return (
+        valid &&
+        validateField(validatedResult[key], untrustedDataTransformed[key])
+      );
+    }
+    return false;
+  }, true);
+
+  return isValid;
+};
+
+export const validateFramesPost = async (
+  payload: DscvrFramesRequest,
+  apiUrl = DEFAULT_DSCVR_API_URL
+): Promise<DscvrValidationResponse> => {
+  const client = new Client({
+    url: apiUrl,
+    exchanges: [cacheExchange, fetchExchange],
+  });
```

Comment:
maybe we create a Client in `src/clients/index.ts` and use it like

```js
const { client } = useUrqlClient({ url: apiUrl }); 
```

or even add `apiUrl` in `clients/index.ts` and just do `const { client } = useUrqlClient();`

again, not a blocker, just a suggestion.

Reply:
Done